### PR TITLE
[ActivityIndicator] Default cycleColors property for empty arrays #1508

### PR DESCRIPTION
**Pull request by romoore**
The `cycleColors` property currently handles empty array assignments by assigning an array with a
single blue color.  This change will modify that behavior to revert back to the original "default"
array.

* Adding unit tests to confirm the default behavior in different scenarios.

### CODE REVIEW

File: components/ActivityIndicator/src/MDCActivityIndicator.h
Comment (jverkoey) [inline components/ActivityIndicator/src/MDCActivityIndicator.h]: A suggestion: "If an empty array is provided to this property's setter, then the provided array will be discarded and an array consisting of the default color values will be assigned instead."

**Comment by jverkoey** [comment]
Two independent topics of feedback:

## On client considerations

Looking at this PR from the client's point of view, I worry that lines like this will require client-side documentation in order to make it clear what's happening:

```
// Reset the color to the default
indicator.cycleColors = @[];
```

If, instead, we provided an `extern NSArray *const MDCActivityIndicatorDefaultCycleColors` that can be assigned instead, I think we'd find that client code will be more self-documenting:

```
indicator.cycleColors = MDCActivityIndicatorDefaultCycleColors;
```

## On the meaning of an empty array

I do appreciate that this PR as is resolves the meaning of an "empty" array of colors for the activity indicator. The above suggestion re-introduces this problem. But I don't think that defaulting to a blue color is much better.

One suggestion for resolving the empty array problem would be to treat it literally and hide the activity indicator if there are no colors to be shown.

**Review by jverkoey — CHANGES_REQUESTED** [review]

File: components/ActivityIndicator/src/MDCActivityIndicator.m
Comment (ianegordon) [inline components/ActivityIndicator/src/MDCActivityIndicator.m]: Simplify as to defaultCycleColors as discussed offline.

File: components/ActivityIndicator/src/MDCActivityIndicator.m
Comment (ianegordon) [inline components/ActivityIndicator/src/MDCActivityIndicator.m]: Simplify to just returning _ivar as discussed offline.

File: components/ActivityIndicator/src/MDCActivityIndicator.m
Comment (ianegordon) [inline components/ActivityIndicator/src/MDCActivityIndicator.m:412]: Based on _cycleColors never being empty, consider simplifing to self.strokeColor = self.cycleColors[currentColorCount];

File: components/ActivityIndicator/src/MDCActivityIndicator.m
Comment (ianegordon) [inline components/ActivityIndicator/src/MDCActivityIndicator.m]: Assuming _cycleColors is never empty:
_currentColorCount = (currentColorCount + 1) % self.cycleColors.count;
Consider s/_currentColorCount/_currentColorIndex/

File: components/ActivityIndicator/src/MDCActivityIndicator.h
Comment (ianegordon) [reply components/ActivityIndicator/src/MDCActivityIndicator.h]: +1 to Jeff's suggestion.

**Review by ianegordon — CHANGES_REQUESTED** [review]

File: components/ActivityIndicator/src/MDCActivityIndicator.h
Comment (romoore) [reply components/ActivityIndicator/src/MDCActivityIndicator.h]: Done

File: components/ActivityIndicator/src/MDCActivityIndicator.m
Comment (romoore) [reply components/ActivityIndicator/src/MDCActivityIndicator.m]: Done

File: components/ActivityIndicator/src/MDCActivityIndicator.m
Comment (romoore) [reply components/ActivityIndicator/src/MDCActivityIndicator.m]: Done

File: components/ActivityIndicator/src/MDCActivityIndicator.m
Comment (romoore) [reply components/ActivityIndicator/src/MDCActivityIndicator.m:412]: Currently the only callers are from `-startAnimating` which sets it to 0, and `-strokeRotationCycleFinishedFromState:` which does an increment/wrap-around.  You want to remove the bounds check and depend on the caller having first reset the index?

**Comment by romoore** [comment]
@featherless @ianegordon  Patch is up and I believe I've addressed your comments. Please take a look when you have time.

File: components/ActivityIndicator/tests/unit/ActivityIndicatorTests.m
Comment (jverkoey) [inline components/ActivityIndicator/tests/unit/ActivityIndicatorTests.m]: `cycleColors` here and throughout.

File: components/ActivityIndicator/tests/unit/ActivityIndicatorTests.m
Comment (jverkoey) [inline components/ActivityIndicator/tests/unit/ActivityIndicatorTests.m]: Do you need to reset this in the test in order to avoid side effects?

**Review by jverkoey — APPROVED** [review]
Looks good! Just a couple questions about the tests.

**Comment by ianegordon** [comment]
Closes #1508

File: components/ActivityIndicator/src/MDCActivityIndicator.m
Comment (willlarche) [inline components/ActivityIndicator/src/MDCActivityIndicator.m:792]: Can we not get the colors from MDCPalette?

File: components/ActivityIndicator/src/MDCActivityIndicator.h
Comment (willlarche) [inline components/ActivityIndicator/src/MDCActivityIndicator.h:105]: I'm gonna say that nonnull feels odd here. I understand wanting to support the empty array. Good idea. But sending empty collections to reset the collection returned is unusual.

Comment (willlarche) [reply components/ActivityIndicator/src/MDCActivityIndicator.h:105]: So I would expect null_resettable. But I know it's UIAppearance compatible. So nullable then.

File: components/ActivityIndicator/src/MDCActivityIndicator.m
Comment (willlarche) [inline components/ActivityIndicator/src/MDCActivityIndicator.m:787]: We also do this stuff with static functions. Would be good to have a best practice defined.

File: components/ActivityIndicator/tests/unit/ActivityIndicatorTests.m
Comment (willlarche) [inline components/ActivityIndicator/tests/unit/ActivityIndicatorTests.m:67]: Would be good, since nullability is involved, to add a swift test too.

**Review by willlarche — CHANGES_REQUESTED** [review]

**Review by ianegordon — APPROVED** [review]

File: components/ActivityIndicator/tests/unit/ActivityIndicatorTests.m
Comment (romoore) [reply components/ActivityIndicator/tests/unit/ActivityIndicatorTests.m]: Good catch, fixed.

File: components/ActivityIndicator/tests/unit/ActivityIndicatorTests.m
Comment (romoore) [reply components/ActivityIndicator/tests/unit/ActivityIndicatorTests.m]: Yep, I think so.  Although the proxy won't actually do anything in the Unit Test target (because there's never a displayed view so no view hierarchy), this will definitely result in a persistent change to the testing environment.  I'm going to remove this test since it's not actually useful.

File: components/ActivityIndicator/src/MDCActivityIndicator.m
Comment (romoore) [reply components/ActivityIndicator/src/MDCActivityIndicator.m:792]: I believe that would be outside the scope of this particular issue, and I opened Issue #1546 to track the request.

File: components/ActivityIndicator/src/MDCActivityIndicator.m
Comment (romoore) [reply components/ActivityIndicator/src/MDCActivityIndicator.m:787]: In this case the current implementation has no C-style functions and it does have `+ (CGFloat)defaultHeight`, so I was matching that style.  Looking at the other components, I see we `static inline` for single objects and primitives/structs, but no others for a collection like this.  I'd like to leave it as-is for now and discuss changing it in the future.

File: components/ActivityIndicator/src/MDCActivityIndicator.h
Comment (romoore) [reply components/ActivityIndicator/src/MDCActivityIndicator.h:105]: I agree that it's a bit unusual for a UI component.  Unfortunately it would be a pretty big change for clients who are depending on the current nonnull attributes.  With your blessing I'd like to leave it as-is for now.

**Review by willlarche — APPROVED** [review]

**Comment by willlarche** [comment]
Approved. Waiting on CI green light.

**Pull request merged**
